### ISSUE TITLE
usingFastIndex: indexes may be created while restore is in progress

### ISSUE THREAD

**Issue by SeanTobin** [maxbube/mydumper#292]
During a DB restore using usingFastIndex, I encountered an issue where we had one (very large) table restore at the end still running. Looking back through the logs, it appeared that the indices were created for this table while the data restore was still in progress. 

The behavior I would expect from usingFastIndex would be to either block index creation of tables until their data restore is complete, or block index creation of all tables until all the data restores are complete.

**Comment by davidducos** [comment]
Hi @SeanTobin,

Did you use --chunk-filesize or --rows ? take into account that if you don't use this parameters the single insert file will use just one thread or if the table doesn't have an integer primary key.

**Comment by SeanTobin** [comment]
We would have liked to use either --chunk-filesize or --rows, but I keep ending up with deadlocks when using them. Had to use a single insert per table in order to get the data restore to complete.

**Comment by davidducos** [comment]
@SeanTobin,
Ok, so, this is a feature request, not a bug. We could add a the mechanism that you describe, but we will need to keep track of the amount of INSERT files applied per table or wait until the queue gets empty to enqueue the ALTER TABLE ADD INDEX statements.
The first option is better but more complex, and it might give us more opportunities in the future.
I need to think this a bit more. 
It is weird to hear that you have deadlocks, can you create another issue with details to reproduce your case? it will be really useful to build our testing environment to avoid bugs on new releases.

**Comment by davidducos** [comment]
@SeanTobin 
I thought about this, and the only case where adding a sync method will be if you have a single large table and it is not being split into chunks, and that is a wrong usage of MyDumper. 
For instance, if you import with 4 threads and have a table with 1M rows and you are using -r 10k, you will end up with 100 data files. During the import, the ALTER TABLE that adds the indexes will happen when just 4 data files remains that are going to finish fast and the alter will continue running. 
Please, let me know what is your use case that you need to sync before continue.

**Comment by SeanTobin** [comment]
@davidducos - Yes, that's the scenario that I'm working with. I have one table that is several times larger than its peers.

**Comment by davidducos** [comment]
@SeanTobin,

Ok, I will be adding that feature but not by default, you will need to use a parameter like --sync-before-add-index. 

I will need to move the queue with the alter tables to wider scope and for the sync I will need to add a waiting queue where the process thread will stop when a new job type JOB_WAIT reaches. 

Again, I don't like it, but I can see the use case. BTW could be possible to give me more information about the deadlock issue? I think that it will be really useful for the community if we can sort it out.

**Comment by davidducos** [comment]
@SeanTobin I added the --sync-before-add-index parameter to the pull request... feel free to test it.

**Comment by SeanTobin** [comment]
Just completed a restore using --sync-before-add-index and it worked as expected

**Comment by davidducos** [comment]
Excellent!

**Comment by davidducos** [comment]
I found that the conde is not using --sync-before-add-index, some changes in the code was done that might help, but the expected intention is not there. I'll be working on it before releasing 0.10.7

**Comment by matthewlenz** [comment]
Is this issue resolved?  If so, I just experienced this issue running the newest version as of today.

**Comment by bato3** [comment]
@matthewlenz Can you provide more info?

**Comment by matthewlenz** [comment]
@bato3 db is about 30GB on disk.  I used most of the standard mydumper options so the export for each table was a single file.  myloader was run with standard options plus --innodb-optimize-keys and --sync-before-add-index.

The table I saw this happen on is the largest in the entire DB with about 80 million rows.  It was still inserting rows long after all the other threads had finished insert and alter.  I happened to notice in the process list that the ALTER to add the keys on this table was being executed while the data inserts on that same table were still happening.

**Comment by davidducos** [comment]
@matthewlenz you might be using version 0.10.7-1 which was not the same that @bato3 has been testing. Please, use master or wait until 0.10.9 has been released as they have a different implementation.

**Comment by matthewlenz** [comment]
@davidducos, I'm using the newest released package mydumper_0.10.7-2.buster_amd64.deb

**Comment by davidducos** [comment]
@matthewlenz ok, that explains it, 0.10.7-2 has a buggy implementation of --innodb-optimize-keys. Use master or wait until 0.10.9.

**Comment by matthewlenz** [comment]
@davidducos will do.  I do have one question.  When this was happening I ran into a different related issue with innodb_online_alter_log_max_size being exceeded during the alter phase on that same table.  Do you think that is a side effect of this broken --innodb-optimize-keys/--sync-before-add-index implementation?  If not do you think it would be worthwhile to attempt some kind of calculation to determine if the default size of that variable needs to be increased temporarily by myloader?

**Comment by davidducos** [comment]
That is a side effect of this implementation. @bato3 knows about it as he helped me a lot troubleshooting all those issues. I thought that the implementation of --innodb-optimize-keys that we released on v0.10.7 was good enough, but it was not. I tried to fix all those issues and now I feel more confortable with new implementation. I'm not saying that it bug free, what I mean is that it has less bugs and we are going to be able to fix them much faster.
